### PR TITLE
feat: Skinning FileSystem

### PR DESCRIPTION
**Pull request by sncarbo**

### CODE REVIEW

**Review by DADSILVER — COMMENTED** [review]
Good job, but let me tell you some suggestions.

File: Source/DataModels/Resources/ResourceMesh.h [hunk 1]
```diff
@@ -63,6 +91,8 @@ class ResourceMesh : virtual public Resource
 	void CreateEBO();
 	void CreateVAO();
 
+	static constexpr unsigned int bonesPerVertex = 4;
```
[the remark this answers is not in the document]
Comment (sncarbo) [reply Source/DataModels/Resources/ResourceMesh.h:96]: Done! Thank you for noticing.

**Review by DADSILVER — APPROVED** [review]

File: Source/DataModels/Resources/ResourceMesh.h [hunk 2]
```diff
@@ -33,24 +48,37 @@ class ResourceMesh : virtual public Resource
 	unsigned int GetNumVertices();
 	unsigned int GetNumFaces();
 	unsigned int GetNumIndexes();
+	unsigned int GetNumBones();
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Resources/ResourceMesh.h:51]: I think this and the rest of the getters of this class can be const

File: Source/DataModels/Resources/ResourceMesh.h [hunk 3]
```diff
@@ -146,6 +187,21 @@ inline const std::vector<std::vector<unsigned int> >& ResourceMesh::GetFacesIndi
...
+}
+
+inline const std::vector<Attach>& ResourceMesh::GetAttaches() const 
+{
+	return attaches;
+}
+
+inline const std::vector<unsigned int>& ResourceMesh::GetNumWeights() const
+{
+	return this->numWeights;
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Resources/ResourceMesh.h:202]: can we do it this way? If it works I think is more simple
```suggestion
	return numWeights;
```

File: Source/DataModels/Resources/ResourceMesh.h [hunk 4]
```diff
@@ -196,3 +257,36 @@ inline void ResourceMesh::SetFacesIndices(const std::vector<std::vector<unsigned
...
 }
+
+inline void ResourceMesh::SetBones(const std::vector<Bone>& bones)
+{
+	this->bones = bones;
+}
+
+inline void ResourceMesh::SetAttachResize()
+{
+	this->attaches.resize(this->numVertices);
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Resources/ResourceMesh.h:268]: do you think we can avoid using `this->` to simplify this kind of methods?

File: Source/FileSystem/Importers/MeshImporter.cpp [hunk 5]
```diff
@@ -166,5 +213,50 @@ void MeshImporter::Load(const char* fileBuffer, std::shared_ptr<ResourceMesh> re
...
+
+	std::vector<Bone> bones;
+	std::vector<unsigned int> allNumWeights;
+	resource->SetAttachResize();
+	unsigned int vertexId = 0u;
+	float vertexWeight = 0.0f;
+	unsigned int numWeights = 0u;
+	for (unsigned int i = 0; i < resource->GetNumBones(); ++i)
+	{
+		Bone bone;
```
Comment (MarcGonzalezMoratona) [inline Source/FileSystem/Importers/MeshImporter.cpp:228]: if this bone is only used in the scope of this loop, is there a way to delete its memory once we have finished working with it?

File: Source/DataModels/Resources/ResourceMesh.h (hunk 2 above)
Comment (sncarbo) [reply Source/DataModels/Resources/ResourceMesh.h:51]: Done! Thank you for pointing this out.

File: Source/DataModels/Resources/ResourceMesh.h (hunk 3 above)
Comment (sncarbo) [reply Source/DataModels/Resources/ResourceMesh.h:202]: I think that according to the programming rules it has to be in this format.

**Review by MarcGonzalezMoratona — DISMISSED** [review]

File: Source/DataModels/Resources/ResourceMesh.h [hunk 6]
```diff
@@ -71,12 +103,16 @@ class ResourceMesh : virtual public Resource
...
 	unsigned int numIndexes;
+	unsigned int numBones;
 	unsigned int materialIndex;
 	std::vector<float3> vertices;
 	std::vector<float3> textureCoords;
 	std::vector<float3> normals;
 	std::vector<float3> tangents{};
 	std::vector<std::vector<unsigned int> > facesIndices;
+	std::vector<Bone> bones;
+	std::vector<Attach> attaches;
```
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:114]: Why don't u make these vectors a vector of pointers? will be easier to pass them through the getters

File: Source/DataModels/Resources/ResourceMesh.h [hunk 7]
```diff
@@ -116,36 +152,56 @@ inline unsigned int ResourceMesh::GetNumIndexes()
...
+}
+
+inline const std::vector<Attach>& ResourceMesh::GetAttaches() const 
+{
+	return attaches;
+}
+
+inline const std::vector<unsigned int>& ResourceMesh::GetNumWeights() const
+{
+	return this->numWeights;
```
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:202]: this-> is not necessary

File: Source/DataModels/Resources/ResourceMesh.h (hunk 4 above)
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:268]: this-> is not necessary in both variables

File: Source/DataModels/Resources/ResourceMesh.h [hunk 8]
```diff
@@ -196,3 +257,36 @@ inline void ResourceMesh::SetFacesIndices(const std::vector<std::vector<unsigned
...
+}
+
+inline void ResourceMesh::SetAttachResize()
+{
+	this->attaches.resize(this->numVertices);
+}
+
+inline void ResourceMesh::SetAttachBones(const unsigned int vertexId, const unsigned int boneId)
+{
+	this->attaches[vertexId].bones[this->attaches[vertexId].numBones] = boneId;
```
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:273]: this-> is not necessary

File: Source/DataModels/Resources/ResourceMesh.h [hunk 9]
```diff
@@ -196,3 +257,36 @@ inline void ResourceMesh::SetFacesIndices(const std::vector<std::vector<unsigned
...
+inline void ResourceMesh::SetNumWeights(const std::vector<unsigned int>& numWeights)
+{
+	this->numWeights = numWeights;
+}
+
+inline void ResourceMesh::IncrementAttachNumBones(const unsigned int vertexId)
+{
+	if (this->attaches[vertexId].numBones < bonesPerVertex)
+	{
+		++this->attaches[vertexId].numBones;
```
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:290]: this-> is not necessary

File: Source/DataModels/Resources/ResourceMesh.h [hunk 10]
```diff
@@ -196,3 +257,36 @@ inline void ResourceMesh::SetFacesIndices(const std::vector<std::vector<unsigned
...
+}
+
+inline void ResourceMesh::SetNumWeights(const std::vector<unsigned int>& numWeights)
+{
+	this->numWeights = numWeights;
+}
+
+inline void ResourceMesh::IncrementAttachNumBones(const unsigned int vertexId)
+{
+	if (this->attaches[vertexId].numBones < bonesPerVertex)
```
Comment (PaablooCH) [inline Source/DataModels/Resources/ResourceMesh.h:288]: this-> is not necessary

**Review by PaablooCH — CHANGES_REQUESTED** [review]
Hi, good job. I see some things in the code:

File: Source/DataModels/Resources/ResourceMesh.h (hunk 10 above)
Comment (sncarbo) [reply Source/DataModels/Resources/ResourceMesh.h]: Done! I have removed all the `this` that you mentioned above.

File: Source/DataModels/Resources/ResourceMesh.h (hunk 6 above)
Comment (sncarbo) [reply Source/DataModels/Resources/ResourceMesh.h:114]: The struct destructors need to be explicitly defined and I had thought that this was not necessary as the size in bytes should not be very large.

**Review by PaablooCH — APPROVED** [review]

**Review by MacXxs — APPROVED** [review]
Everything working! 👍🏼

**Pull request merged**